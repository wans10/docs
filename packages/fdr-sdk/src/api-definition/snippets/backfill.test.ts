import { describe, expect, it } from "vitest";

import {
  ApiDefinition,
  ApiDefinitionId,
  EndpointId,
  EnvironmentId,
} from "../latest";
import { backfillSnippets } from "./backfill";

describe("backfillSnippets", () => {
  it("should backfill snippets for a search endpoint example", async () => {
    const apiDefinition: ApiDefinition = {
      id: ApiDefinitionId("test-api"),
      endpoints: {
        [EndpointId("search")]: {
          id: EndpointId("search"),
          method: "POST",
          path: [{ type: "literal", value: "/" }],
          displayName: undefined,
          operationId: undefined,
          auth: undefined,
          defaultEnvironment: undefined,
          environments: [
            {
              id: EnvironmentId("default"),
              baseUrl: "https://api.example.com/v1",
            },
          ],
          pathParameters: undefined,
          queryParameters: undefined,
          requestHeaders: undefined,
          responseHeaders: undefined,
          requests: undefined,
          responses: undefined,
          errors: undefined,
          snippetTemplates: undefined,
          protocol: undefined,
          description: undefined,
          availability: undefined,
          namespace: undefined,
          examples: [
            {
              name: "Basic Search",
              description: "",
              path: "/",
              pathParameters: {},
              queryParameters: {},
              headers: {},
              requestBody: {
                type: "json",
                value: {
                  RecordCount: 50,
                  DocumentSearchParams: {
                    SearchTerms: {
                      All: ["data security"],
                      Any: ["cyberattack", "breach"],
                      None: ["ransomware"],
                    },
                    DocumentDateRangeStart: "2023-01-01T00:00:00Z",
                    DocumentDateRangeEnd: "2023-12-31T23:59:59Z",
                  },
                },
              },
              responseStatusCode: 200,
              responseBody: {
                type: "json",
                value: {
                  TotalDocumentCount: 833494,
                  NotReturnedDocumentCount: 0,
                  ReturnedDocumentCount: 2,
                  TruncatedDocumentCount: 0,
                  Documents: [
                    {
                      IsTruncated: false,
                      DocumentID: 77187402516,
                      DocumentType: "NEWS",
                      UploadDate: "2024-10-22T16:54:36Z",
                      DocumentDate: "2024-10-22T15:17:54Z",
                      DocumentName:
                        "Wall Street Continues to See Recent Rally Cooling",
                      DocumentText: "U.S. stocks dropped again on Tuesday...",
                      AuthorName: "Natalie Venegas",
                      DocumentSourceURL:
                        "https://www.newsweek.com/wall-street-continues-see-recent-rally-cooling-1972948",
                      DocumentImageURL:
                        "https://d.newsweek.com/en/full/2501402/new-york-stock-exchange.jpg",
                      LanguageID: "en",
                      DocumentSentimentScore: "-13",
                      ContainsViolence: true,
                    },
                  ],
                  FavIcons: {
                    NEWS: "https://d1hgo075dbsz4i.cloudfront.net/21f4f43585ea45aa539034866e692e21/a/News/dsicon",
                  },
                },
              },
              snippets: undefined,
            },
          ],
        },
      },
      auths: {},
      websockets: {},
      webhooks: {},
      types: {},
      globalHeaders: [],
      subpackages: {},
      snippetsConfiguration: undefined,
    };

    const flags = {
      usesApplicationJsonInFormDataValue: false,
      isHttpSnippetsEnabled: true,
      alwaysEnableJavaScriptFetch: true,
    };

    const result = await backfillSnippets(apiDefinition, flags);

    // Verify the result has the expected structure
    const endpoint = result.endpoints[EndpointId("search")];
    expect(endpoint).toBeDefined();
    const examples = endpoint?.examples;
    expect(examples).toHaveLength(1);

    const example = examples?.[0];
    expect(example).toBeDefined();

    // Verify snippets were generated
    const snippets = example?.snippets;
    expect(snippets).toBeDefined();
    const curlSnippets = snippets?.curl;
    expect(curlSnippets).toBeDefined();
    expect(curlSnippets).toHaveLength(1);

    // Verify curl snippet contains expected content
    const curlSnippet = curlSnippets?.[0];
    expect(curlSnippet?.language).toBe("curl");
    expect(curlSnippet?.code).toContain("curl");
    expect(curlSnippet?.code).toContain("POST");
    expect(curlSnippet?.code).toContain('"RecordCount": 50');
    expect(curlSnippet?.code).toContain('"data security"');

    expect(example?.snippets).toMatchSnapshot();
  });
});
