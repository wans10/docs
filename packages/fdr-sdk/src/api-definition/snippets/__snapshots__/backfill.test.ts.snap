// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`backfillSnippets > should backfill snippets for a search endpoint example 1`] = `
{
  "csharp": [
    {
      "code": "var client = new RestClient("https://api.example.com/v1/");
var request = new RestRequest(Method.POST);
request.AddHeader("Content-Type", "application/json");
request.AddParameter("application/json", "{\\n  \\"RecordCount\\": 50,\\n  \\"DocumentSearchParams\\": {\\n    \\"SearchTerms\\": {\\n      \\"All\\": [\\n        \\"data security\\"\\n      ],\\n      \\"Any\\": [\\n        \\"cyberattack\\",\\n        \\"breach\\"\\n      ],\\n      \\"None\\": [\\n        \\"ransomware\\"\\n      ]\\n    },\\n    \\"DocumentDateRangeStart\\": \\"2023-01-01T00:00:00Z\\",\\n    \\"DocumentDateRangeEnd\\": \\"2023-12-31T23:59:59Z\\"\\n  }\\n}", ParameterType.RequestBody);
IRestResponse response = client.Execute(request);",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "csharp",
      "name": undefined,
    },
  ],
  "curl": [
    {
      "code": "curl -X POST https://api.example.com/v1/ \\
     -H "Content-Type: application/json" \\
     -d '{
  "RecordCount": 50,
  "DocumentSearchParams": {
    "SearchTerms": {
      "All": [
        "data security"
      ],
      "Any": [
        "cyberattack",
        "breach"
      ],
      "None": [
        "ransomware"
      ]
    },
    "DocumentDateRangeStart": "2023-01-01T00:00:00Z",
    "DocumentDateRangeEnd": "2023-12-31T23:59:59Z"
  }
}'",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "curl",
      "name": undefined,
    },
  ],
  "go": [
    {
      "code": "package main

import (
	"fmt"
	"strings"
	"net/http"
	"io"
)

func main() {

	url := "https://api.example.com/v1/"

	payload := strings.NewReader("{\\n  \\"RecordCount\\": 50,\\n  \\"DocumentSearchParams\\": {\\n    \\"SearchTerms\\": {\\n      \\"All\\": [\\n        \\"data security\\"\\n      ],\\n      \\"Any\\": [\\n        \\"cyberattack\\",\\n        \\"breach\\"\\n      ],\\n      \\"None\\": [\\n        \\"ransomware\\"\\n      ]\\n    },\\n    \\"DocumentDateRangeStart\\": \\"2023-01-01T00:00:00Z\\",\\n    \\"DocumentDateRangeEnd\\": \\"2023-12-31T23:59:59Z\\"\\n  }\\n}")

	req, _ := http.NewRequest("POST", url, payload)

	req.Header.Add("Content-Type", "application/json")

	res, _ := http.DefaultClient.Do(req)

	defer res.Body.Close()
	body, _ := io.ReadAll(res.Body)

	fmt.Println(res)
	fmt.Println(string(body))

}",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "go",
      "name": undefined,
    },
  ],
  "java": [
    {
      "code": "HttpResponse<String> response = Unirest.post("https://api.example.com/v1/")
  .header("Content-Type", "application/json")
  .body("{\\n  \\"RecordCount\\": 50,\\n  \\"DocumentSearchParams\\": {\\n    \\"SearchTerms\\": {\\n      \\"All\\": [\\n        \\"data security\\"\\n      ],\\n      \\"Any\\": [\\n        \\"cyberattack\\",\\n        \\"breach\\"\\n      ],\\n      \\"None\\": [\\n        \\"ransomware\\"\\n      ]\\n    },\\n    \\"DocumentDateRangeStart\\": \\"2023-01-01T00:00:00Z\\",\\n    \\"DocumentDateRangeEnd\\": \\"2023-12-31T23:59:59Z\\"\\n  }\\n}")
  .asString();",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "java",
      "name": undefined,
    },
  ],
  "javascript": [
    {
      "code": "const url = 'https://api.example.com/v1/';
const options = {
  method: 'POST',
  headers: {'Content-Type': 'application/json'},
  body: '{"RecordCount":50,"DocumentSearchParams":{"SearchTerms":{"All":["data security"],"Any":["cyberattack","breach"],"None":["ransomware"]},"DocumentDateRangeStart":"2023-01-01T00:00:00Z","DocumentDateRangeEnd":"2023-12-31T23:59:59Z"}}'
};

try {
  const response = await fetch(url, options);
  const data = await response.json();
  console.log(data);
} catch (error) {
  console.error(error);
}",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "javascript",
      "name": undefined,
    },
  ],
  "php": [
    {
      "code": "<?php

$client = new \\GuzzleHttp\\Client();

$response = $client->request('POST', 'https://api.example.com/v1/', [
  'body' => '{
  "RecordCount": 50,
  "DocumentSearchParams": {
    "SearchTerms": {
      "All": [
        "data security"
      ],
      "Any": [
        "cyberattack",
        "breach"
      ],
      "None": [
        "ransomware"
      ]
    },
    "DocumentDateRangeStart": "2023-01-01T00:00:00Z",
    "DocumentDateRangeEnd": "2023-12-31T23:59:59Z"
  }
}',
  'headers' => [
    'Content-Type' => 'application/json',
  ],
]);

echo $response->getBody();",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "php",
      "name": undefined,
    },
  ],
  "python": [
    {
      "code": "import requests

url = "https://api.example.com/v1/"

payload = {
    "RecordCount": 50,
    "DocumentSearchParams": {
        "SearchTerms": {
            "All": ["data security"],
            "Any": ["cyberattack", "breach"],
            "None": ["ransomware"]
        },
        "DocumentDateRangeStart": "2023-01-01T00:00:00Z",
        "DocumentDateRangeEnd": "2023-12-31T23:59:59Z"
    }
}
headers = {"Content-Type": "application/json"}

response = requests.post(url, json=payload, headers=headers)

print(response.json())",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "python",
      "name": undefined,
    },
  ],
  "ruby": [
    {
      "code": "require 'uri'
require 'net/http'

url = URI("https://api.example.com/v1/")

http = Net::HTTP.new(url.host, url.port)
http.use_ssl = true

request = Net::HTTP::Post.new(url)
request["Content-Type"] = 'application/json'
request.body = "{\\n  \\"RecordCount\\": 50,\\n  \\"DocumentSearchParams\\": {\\n    \\"SearchTerms\\": {\\n      \\"All\\": [\\n        \\"data security\\"\\n      ],\\n      \\"Any\\": [\\n        \\"cyberattack\\",\\n        \\"breach\\"\\n      ],\\n      \\"None\\": [\\n        \\"ransomware\\"\\n      ]\\n    },\\n    \\"DocumentDateRangeStart\\": \\"2023-01-01T00:00:00Z\\",\\n    \\"DocumentDateRangeEnd\\": \\"2023-12-31T23:59:59Z\\"\\n  }\\n}"

response = http.request(request)
puts response.read_body",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "ruby",
      "name": undefined,
    },
  ],
  "swift": [
    {
      "code": "import Foundation

let headers = ["Content-Type": "application/json"]
let parameters = [
  "RecordCount": 50,
  "DocumentSearchParams": [
    "SearchTerms": [
      "All": ["data security"],
      "Any": ["cyberattack", "breach"],
      "None": ["ransomware"]
    ],
    "DocumentDateRangeStart": "2023-01-01T00:00:00Z",
    "DocumentDateRangeEnd": "2023-12-31T23:59:59Z"
  ]
] as [String : Any]

let postData = JSONSerialization.data(withJSONObject: parameters, options: [])

let request = NSMutableURLRequest(url: NSURL(string: "https://api.example.com/v1/")! as URL,
                                        cachePolicy: .useProtocolCachePolicy,
                                    timeoutInterval: 10.0)
request.httpMethod = "POST"
request.allHTTPHeaderFields = headers
request.httpBody = postData as Data

let session = URLSession.shared
let dataTask = session.dataTask(with: request as URLRequest, completionHandler: { (data, response, error) -> Void in
  if (error != nil) {
    print(error as Any)
  } else {
    let httpResponse = response as? HTTPURLResponse
    print(httpResponse)
  }
})

dataTask.resume()",
      "description": undefined,
      "generated": true,
      "install": undefined,
      "language": "swift",
      "name": undefined,
    },
  ],
}
`;
